feat: UserSetting UI 구현 및 입력받은 username 으로 수정기능

parkyounbae opened the pull request:
프로필 사진을 표시하는 이미지뷰
유저이름을 표시하는 텍스트필드
텍스트 필드를 활성화/비활성화 해주는 edit 버튼
입력 완료시 수정된 이름을 저장하는 done 버튼
아무런 행동을 하지 않고 나가는 cancel 버튼 구현

Resolves: GITIT-61

parkyounbae committed:
feat: UserSetting UI 구현 및 입력받은 username 으로 수정기능

프로필 사진을 표시하는 이미지뷰
유저이름을 표시하는 텍스트필드
텍스트 필드를 활성화/비활성화 해주는 edit 버튼
입력 완료시 수정된 이름을 저장하는 done 버튼
아무런 행동을 하지 않고 나가는 cancel 버튼 구현

Resolves: GITIT-61

parkyounbae committed:
Fix: frame과 오토레이아웃이 중복으로 들어가 있는 부분 삭제

frame설정 관련 코드 삭제

Resoves:GITIT-61

sweethoneybee commented:
오토레이아웃도 같이 적용하신다고 고생하셨어요! 몇가지 피드백 드리자면  

1. 클로저로 uiview 세팅하고 리턴값을 새로운 지역변수에 할당하고 있는데, 바로 프로퍼티에 할당해도 되지 않을까요? 
2. `UserInfo.Key.username.rawValue`값으로 `textField.text`를 초기화한 특별한 이유가 있으신가요? `UserInfo.Key`는 `UserInfo`의 메소드 인자에 사용하도록 만들어 둔 거라 클래스 밖에서는 딱히 사용할 일이 없을 거예요. 만약 username이 필요한 거라면 `UserInfo.username`을 사용하시면 됩니다.  
3. username을 편집하는 UX가 조금 어색해요. 지금 구현한 방식은 '편집버튼 누름 -> textField 클릭 -> 편집버튼 다시 누름 or Done 클릭'으로 되어있는데, 보통 우리가 다른 앱을 생각해보면 편집버튼을 누르거나 textField를 누르면 바로 키보드가 올라오잖아요? 그걸 한번 고려해서 UX를 다시 구성해주세요. 키보드를 올라오게 하려면 textField가 firstResponder가 되면 됩니다(`becomeFirstResponder` 메소드를 찾아보세요!).   
4. 지금 편집버튼을 따로 생성하고 위치도 잡아주고 있는데, UITextField의 OverlayView에 대해서 찾아보고 적용해보세요. 구상하셨던 것 그대로를 훨씬 더 정확하게 구현하실 수 있을 거예요. ([https://developer.apple.com/documentation/uikit/uitextfield](https://developer.apple.com/documentation/uikit/uitextfield))
5. done 버튼은 중앙에도 있지만 좌상단 cancel 버튼처럼 우상단에도 done 버튼을 만들어주세요. 대부분 앱을 보면 중앙보다는 우상단에 done만 위치하더라고요. 
6. cancel, done 버튼을 좌상단 우상단에 위치를 잡아줄 필요 없이, `UIViewController.navigationItem.rightBarButtonItem` 프로퍼티를 이용해보세요. 저희 앱 루트 VC가 UINavigationViewController라서 navgationItem도 사용가능할 거예요. 원하시는 방식을 더 정확하게 구현할 수 있을 것 같습니다.  

추가로, 앱이 작동하는 방식을 어떻게 구상할지 헷갈리면 저는 종종 카톡이나 다른 유명한 앱을 보고 참고하면 구상하기 쉽더라고요. 참고하세요~

sweethoneybee commented:
참, 앱 로직을 우선 구현해보고 나중에 위치 한번에 잡을거라 오토레이아웃 정밀하게 안 잡으셔도 괜찮아요.   
지금은 적당한 곳으로 위치 잡으셔도 됩니당

sweethoneybee reviewed (commented):
오토레이아웃도 같이 적용하신다고 고생하셨어요! 몇가지 피드백 드리자면  

1. 클로저로 uiview 세팅하고 리턴값을 새로운 지역변수에 할당하고 있는데, 바로 프로퍼티에 할당해도 되지 않을까요? 
2. `UserInfo.Key.username.rawValue`값으로 `textField.text`를 초기화한 특별한 이유가 있으신가요? `UserInfo.Key`는 `UserInfo`의 메소드 인자에 사용하도록 만들어 둔 거라 클래스 밖에서는 딱히 사용할 일이 없을 거예요. 만약 username이 필요한 거라면 `UserInfo.username`을 사용하시면 됩니다.  
3. username을 편집하는 UX가 조금 어색해요. 지금 구현한 방식은 '편집버튼 누름 -> textField 클릭 -> 편집버튼 다시 누름 or Done 클릭'으로 되어있는데, 보통 우리가 다른 앱을 생각해보면 편집버튼을 누르거나 textField를 누르면 바로 키보드가 올라오잖아요? 그걸 한번 고려해서 UX를 다시 구성해주세요. 키보드를 올라오게 하려면 textField가 firstResponder가 되면 됩니다(`becomeFirstResponder` 메소드를 찾아보세요!).   
4. 지금 편집버튼을 따로 생성하고 위치도 잡아주고 있는데, UITextField의 OverlayView에 대해서 찾아보고 적용해보세요. 구상하셨던 것 그대로를 훨씬 더 정확하게 구현하실 수 있을 거예요. ([https://developer.apple.com/documentation/uikit/uitextfield](https://developer.apple.com/documentation/uikit/uitextfield))
5. done 버튼은 중앙에도 있지만 좌상단 cancel 버튼처럼 우상단에도 done 버튼을 만들어주세요. 대부분 앱을 보면 중앙보다는 우상단에 done만 위치하더라고요. 
6. cancel, done 버튼을 좌상단 우상단에 위치를 잡아줄 필요 없이, `UIViewController.navigationItem.rightBarButtonItem` 프로퍼티를 이용해보세요. 저희 앱 루트 VC가 UINavigationViewController라서 navgationItem도 사용가능할 거예요. 원하시는 방식을 더 정확하게 구현할 수 있을 것 같습니다.  
7. 시언님과 같은 피드백인데요, 사용자가 편집을 완료하면 키보드가 내려가야하는데 키보드 엔터를 눌렀을 때 내려가는 처리를 해주세요! (다른 곳 터치시 키보드 내려가게 하신 건 잘 구현하셨습니다 bb)  
추가로, 앱이 작동하는 방식을 어떻게 구상할지 헷갈리면 저는 종종 카톡이나 다른 유명한 앱을 보고 참고하면 구상하기 쉽더라고요. 참고하세요~

parkyounbae committed:
style: 피드백 반영

클로저 리턴값을 프로퍼티로 바로 할당
userName 편집 버튼을 텍스트필드에 편입, 버튼 클릭 혹은 텍스트필드 클릭시 키보드
키보드 리턴하면 키보드 숨기기
기존 cancel 버튼 삭제 및 네비게이션 바에 cancel, done버튼 추가
userName값 가져올때 UserInfo.username을 사용

Resolve: GITIT-61

parkyounbae committed:
Style: 피드백 관련 수정사항

클로저의 리턴값을 프로퍼티로 변경
userNameField에 편집버튼 편입
버튼 혹은 필드를 누르면 키보드가 올라오게 수정
키보드에서 리턴을 누르면 키보드가 사라지게 수정
네비게이션 바 아이템을 이용해 done, cancel버튼 추가

Resolve: GITIT-61

sweethoneybee reviewed (changes requested):
1. UITextFieldDelegate 메소드 구현한 부분은 `MARK:-`로 구분하는 건 어떨까요? 나중에 코드를 다시 읽을 때 가독성에 더 도움이 될 거예요

2. UIFunction에서 클로저로 뷰 설정 후 리턴하는 식을 잘 적용하셨네요. 39번째 라인에서 로컬 변수 `imageView`를 따로 사용할 이유가 없다면, 다른 뷰 생성 방식이랑 동일하게 프로퍼티에 할당해주는 것도 좋을 것 같아요. 그리고 뷰를 리턴하면서 프로퍼티에 할당하니깐 클로저 안에서 프로퍼티에 할당 안하셔도 될 것 같아요. 지금은 동일한 의미의 줄이 2개씩 들어가 있는 셈이니 수정하는 게 좋을 것 같아요(ex: 60번째 라인과 66번째 라인은 사실상 동일한 의미).  

3. UIBarButtonItem을 왼쪽, 오른쪽에 잘 추가하셨네요. 혹시 직접 실행 후 결과물도 확인하셨나요? `UIBarButtonItem(barButtonSystemItem:target:action:)` 생성자로 생성하실 때 인자로 넘기는 `barButtonSystemItem`에 따라 버튼의 생김새가 달라져요. 지금은 `compose` 형태로 버튼이 만들어졌을 거예요. 아마 의도하신 건 `cancel`, `done` 일 텐데, 스토리보드에서 barButtonItem 직접 넣고 attribute inspector(xcode 오른쪽에서 설정하는 부분)에서 구체적인 형태 확인할 수 있습니다. 참고해서 수정해보시면 좋을 것 같아요. 

4. 이제 메소드 접근제어자를 수정해볼까요? 지금은 특정 메소드 하나에만 private가 붙어있는데, 외부에서 보이면 안되는 메소드들에 적절한 접근제어자(private 등)을 붙여보세요  

5. 현재 구현하신 거에서는 username을 입력받고, done을 누르면 UserInfo.username을 갱신하면서 화면을 pop하고 있어요. 잘 구현하셨습니다. 이제 세부적으로 textField의 몇몇 설정을 코드로 해주시면 될 것 같아요. 지금은 막 오타 수정해주고, 자동으로 첫글자를 대문자로 하고 있어서 이 부분을 수정해야할 것 같아요! 스토리보드로 하면 클릭 해서 해결 될 걸, 코드로 생성하다보니 코드로 한땀한땀 적어줘야합니다 ㅋㅋ.  

6. Github username에 들어갈 수 있는 문자는 알파벳,숫자,언더바입니다. 공백이나 다른 문자는 들어갈 수 없죠. 클라이언트 단에서도 이를 확인하고 입력을 제한할 필요가 있을 거예요. 검사를 어떻게 할 수 있을까요? textField.text를 매번 검사해서 done을 비활성화할 수도 있고, done을 눌렀을 때만 검사해서 다른 문자가 들어가 있으면 입력을 거부할 수도 있을 것 같아요. 방법을 고민해보고 적용해주세요.


Setting View가 거의다 완성 되가네요! 조금만 더 고생해주세요.

parkyounbae committed:
Style: 코드 중복부분 일부 수정

UITextFieldDelegate부분 주석 추가
userProfile부분 클로저에서 프로퍼티로 바로 할당
add함수에서 프로퍼티 할당부분에 중복되는 부분 수정
UIBarButtonItem 스타일 수정
ui부분 func private 추가

Resolve: GITIT-61

parkyounbae committed:
Feat: textField관련 피드백 반영

userName이 기존의 이름과 다르고 빈 칸이 아니여야 done 버튼 활성화
nameTextField에 영문, 숫자, _ 를 제외한 모든 문자 입력 불가능
textField의 자동완성 기능, 자동 대문자 기능 off

Resolve: GITIT-61

sweethoneybee reviewed (changes requested):
네 고생하셨습니다 훌륭하게 요구사항 다 반영했네요 정규표현식으로 사용자 입력값 필터링하는 것도 좋았어요. 정규표현식은 많은 곳에서 사용되니 잘 알아두면 도움이 될 거예요(근데 저도 종종 까먹어서 치트시트 보면서 합니다 ㅋㅋ).  

1. `String.hasCharacters()` 메소드 구현할 때 아마 노란색 경고가 2개 떴을 건데, 그 경고 다 해결해주세요  
2. VC 프로퍼티가 현재 다 implicit unwrapped optional(! 붙이는 것)로 구현되어 있는데, 전부 ?로 바꿔주세요. 잘못해서 프로퍼티가 nil일 때 접근하면 앱이 크래시될 수 있습니다.  

위에 피드백만 수정후 머지하면 될 것 같습니다. 아래는 2개는 머지 후, 시언님 커밋사항도 반영하고 나면 해결해야할 공통 문제예요. 그래서 둘다 커밋사항을 반영한다음 한 명이 담당하도록 하겠습니다. (아래 2개는 시언님께도 같은 피드백이 갈 예정입니다)

1. 지금은 사용자 입장에서는 영어 소문자, 숫자, 언더바만 입력 가능하다는 걸 알 방법이 없어요. 그래서 그것에 대한 피드백이 필요해요(그 외를 입력하면 알림을 주는 등). 그렇지 않으면 앱이 제대로 동작하고 있다고 생각하지 않을 거예요. 이 문제를 어떻게 해결할지 생각해보고 Login View, Setting View에 적용해야 해요(둘 다 지금 로직이 동일하니). 우리가 웹에서 회원가입을 할 때 잘못된 입력을 넣으면 어떻게 알려주는지를 관찰해보면 답을 얻을 수 있을 거예요.  
2. 윤배님과 시언님 둘다 서로 다른 곳에서 String Extension을 구현하고 있는데 둘다 같은 메소드를 구현하고 있어요. 아마 합치면 컴파일에러가 날 것 같아요. 추후에 중복되는 부분을 분리하여 Extension 디렉토리에서 `String.hasCharacters()`를 구현해주세요.

sweethoneybee commented:
실수로 request change로 달았네요 ㅠㅠ. 다시 리뷰 요청하실 필요 없이 경고 해결 커밋후, 스쿼시 + 머지 해주세요. 깃허브 페이지에서 스쿼시 + 머지 할 때 메시지 이쁘게 잘 작성하시면 됩니다!

parkyounbae committed:
fix: hasCharacters 함수의 경고 수정 및 프로퍼티 옵셔널 수정

hasCharacters 함수 내의 경고를 해결하기 위해 NSMakeRange 대신에 NSRange 사용 및 let _ 대신 nil 을 사용
프로퍼티에 !대신 옵셔널? 을 사용 이에 따라 addSubView시에 옵셔널 바인딩 사용
오토레이아웃을 설정하는 코드를 따로 모아 setAutoLayout 함수 새로 생성

Resolves: GITIT-61
